src/pybind/mgr/osd_perf_query/module.py

```diff
@@ -20,15 +22,85 @@ class OSDPerfQuery(MgrModule):
+        ],
+        'performance_counter_descriptors': [
+            'write_ops', 'read_ops', 'write_bytes', 'read_bytes',
+            'write_latency', 'read_latency',
+        ],
+    }
+
+    RBD_IMAGE_ID_QUERY = {
+        'key_descriptor': [
+            {'type': 'pool_id', 'regex': '^.+$'},
+            {'type': 'object_name', 'regex': '^rbd_data\.([^.]+)\.'},
+        ],
+        'performance_counter_descriptors': [
+            'write_ops', 'read_ops', 'write_bytes', 'read_bytes',
+            'write_latency', 'read_latency',
+        ],
+    }
+
+    queries = {}
+
     def handle_command(self, inbuf, cmd):
         if cmd['prefix'] == "osd perf query add":
-            query_id = self.add_osd_perf_query(cmd['query'])
+            if cmd['query'] == 'rbd_image_id':
+                query = self.RBD_IMAGE_ID_QUERY
+            else:
+                query = self.CLIENT_ID_QUERY
+            query_id = self.add_osd_perf_query(query)
+            if query_id is None:
+                return -errno.EINVAL, "", "Invalid query"
+            self.queries[query_id] = query
             return 0, str(query_id), "added query " + cmd['query'] + " with id " + str(query_id)
         elif cmd['prefix'] == "osd perf query remove":
+            if cmd['query_id'] not in self.queries:
+                return -errno.ENOENT, "", "unknown query id " + str(cmd['query_id'])
             self.remove_osd_perf_query(cmd['query_id'])
+            del self.queries[cmd['query_id']]
             return 0, "", "removed query with id " + str(cmd['query_id'])
+        elif cmd['prefix'] == "osd perf counters get":
+            if cmd['query_id'] not in self.queries:
+                return -errno.ENOENT, "", "unknown query id " + str(cmd['query_id'])
+            query = self.queries[cmd['query_id']]
+            res = self.get_osd_perf_counters(cmd['query_id'])
+
+            if query == self.RBD_IMAGE_ID_QUERY:
+                table = prettytable.PrettyTable(
+                    tuple(["pool_id", "rbd image_id"] + query['performance_counter_descriptors']),
+                    hrules=prettytable.FRAME
+                )
+            else:
+                table = prettytable.PrettyTable(
+                    tuple(["client_id"] + query['performance_counter_descriptors']),
+                    hrules=prettytable.FRAME
+                )
+
+            for c in res['counters']:
+                if query == self.RBD_IMAGE_ID_QUERY:
+                    row = [c['k'][0][0], c['k'][1][1]]
+                else:
+                    row = [c['k'][0][0]]
+                counters = c['c']
+                for i in range(len(query['performance_counter_descriptors'])):
+                    if query['performance_counter_descriptors'][i] in ['write_ops', 'read_ops']:
+                        row.append("%d" % counters[i][0])
+                    else:
+                        row.append("%d/%d" % (counters[i][0], counters[i][1]))
```

Comment:
For this pretty-version, it seems like bytes, throughput, and latency should have proper unit suffixes and provide user-friendly values (i.e. 5 MiB and 24 ms).

Reply:
I was considering this module just for testing purposes and as illustration how it is going to be used, but ok, not much work to make it prettier.